File: desktop/src-tauri/src/commands/printer.rs
use log::{info, warn, error};

// ─── USB Printer ───────────────────────────────────────────────────────────────

/// Print ESC/POS bytes to a USB receipt printer.
///
/// Searches for USB devices with class 0x07 (Printer), opens the first one found,
/// and writes the bytes in 4096-byte chunks to the bulk OUT endpoint.
pub fn print_to_usb(bytes: &[u8]) -> Result<(), String> {
    let devices = rusb::devices().map_err(|e| format!("Failed to enumerate USB devices: {}", e))?;

    for device in devices.iter() {
        let desc = match device.device_descriptor() {
            Ok(d) => d,
            Err(_) => continue,
        };

        // Check if this is a printer (class 0x07)
        if desc.class_code() != 0x07 {
            continue;
        }

        info!(
            "Found USB printer: {:04x}:{:04x}",
            desc.vendor_id(),
            desc.product_id()
        );

        let handle = match device.open() {
            Ok(h) => h,
            Err(e) => {
                warn!("Failed to open USB printer: {}", e);
                continue;
            }
        };

        // Find the printer interface
        for interface in device.active_config_descriptor().map_err(|e| e.to_string())?.interfaces() {
            for iface_desc in interface.descriptors() {
                if iface_desc.class_code() == 0x07 {
                    let iface_num = iface_desc.interface_number();

                    // Try to claim the interface
                    match handle.claim_interface(iface_num) {
                        Ok(()) => {
                            info!("Claimed USB interface {}", iface_num);
                        }
                        Err(e) => {
                            warn!("Failed to claim interface {}: {}", iface_num, e);
                            continue;
                        }
                    }

                    // Find the bulk OUT endpoint
                    for ep_desc in iface_desc.endpoint_descriptors() {
                        if ep_desc.direction() == rusb::Direction::Out
                            && ep_desc.transfer_type() == rusb::TransferType::Bulk
                        {
                            let endpoint_addr = ep_desc.address();
                            info!("Writing {} bytes to USB endpoint 0x{:02x}", bytes.len(), endpoint_addr);

                            // Write in 4096-byte chunks
                            for chunk in bytes.chunks(4096) {
                                handle
                                    .write_bulk(endpoint_addr, chunk, std::time::Duration::from_secs(5))
                                    .map_err(|e| format!("USB write failed: {}", e))?;
                            }

                            info!("USB print complete");
                            return Ok(());
                        }
                    }

                    warn!("No suitable USB endpoint found on interface {}", iface_num);
                }
            }
        }
    }

    Err("No USB receipt printer found".to_string())
}

// ─── Serial/Network Printer ────────────────────────────────────────────────────

/// Print ESC/POS bytes to a serial port (COM/USB-serial).
pub fn print_to_serial(port_name: &str, bytes: &[u8]) -> Result<(), String> {
    let mut port = serialport::new(port_name, 9600)
        .timeout(std::time::Duration::from_secs(5))
        .open()
        .map_err(|e| format!("Failed to open serial port {}: {}", port_name, e))?;

    port.write_all(bytes)
        .map_err(|e| format!("Serial write failed: {}", e))?;

    // Wait for data to be transmitted
    std::thread::sleep(std::time::Duration::from_millis(100));

    info!("Serial print complete on {}", port_name);
    Ok(())
}

/// Print ESC/POS bytes to a network printer via TCP.
pub fn print_to_network(ip: &str, port: u16, bytes: &[u8]) -> Result<(), String> {
    use std::io::Write;
    use std::net::TcpStream;

    let addr = format!("{}:{}", ip, port);
    let mut stream = TcpStream::connect(&addr)
        .map_err(|e| format!("Failed to connect to network printer at {}: {}", addr, e))?;

    stream
        .set_write_timeout(Some(std::time::Duration::from_secs(5)))
        .map_err(|e| e.to_string())?;

    stream
        .write_all(bytes)
        .map_err(|e| format!("Network write failed: {}", e))?;

    stream.flush().map_err(|e| e.to_string())?;

    info!("Network print complete to {}", addr);
    Ok(())
}

// ─── Printer Enumeration ───────────────────────────────────────────────────────

#[derive(Debug, Clone, serde::Serialize)]
pub struct PrinterInfo {
    pub name: String,
    pub port_type: String, // "usb", "serial", "network"
    pub vendor_id: Option<u16>,
    pub product_id: Option<u16>,
    pub port_name: Option<String>,
    pub connected: bool,
}

/// List all available USB printers.
pub fn list_usb_printers() -> Vec<PrinterInfo> {
    let mut printers = Vec::new();

    if let Ok(devices) = rusb::devices() {
        for device in devices.iter() {
            if let Ok(desc) = device.device_descriptor() {
                if desc.class_code() == 0x07 {
                    let name = device
                        .open()
                        .ok()
                        .and_then(|h| h.read_product_string_ascii().ok())
                        .unwrap_or_else(|| format!("USB Printer {:04x}:{:04x}", desc.vendor_id(), desc.product_id()));

                    printers.push(PrinterInfo {
                        name,
                        port_type: "usb".to_string(),
                        vendor_id: Some(desc.vendor_id()),
                        product_id: Some(desc.product_id()),
                        port_name: None,
                        connected: true,
                    });
                }
            }
        }
    }

    printers
}

/// List all available serial ports.
pub fn list_serial_printers() -> Vec<PrinterInfo> {
    let mut printers = Vec::new();

    if let Ok(ports) = serialport::available_ports() {
        for port_info in ports {
            printers.push(PrinterInfo {
                name: port_info.port_name.clone(),
                port_type: "serial".to_string(),
                vendor_id: port_info.vid_pid.map(|v| v.0),
                product_id: port_info.vid_pid.map(|v| v.1),
                port_name: Some(port_info.port_name),
                connected: true,
            });
        }
    }

    printers
}

/// List all printers (USB + Serial).
pub fn list_all_printers() -> Vec<PrinterInfo> {
    let mut printers = list_usb_printers();
    printers.extend(list_serial_printers());
    printers
}

// ─── Cash Drawer ───────────────────────────────────────────────────────────────

/// Send ESC/POS command to open cash drawer via USB printer.
pub fn open_drawer_usb() -> Result<(), String> {
    // ESC/POS: ESC p 0x00 0x19 0xFA — Pulse Pin 2 for 60ms
    let drawer_cmd = vec![0x1B, 0x70, 0x00, 0x19, 0xFA];
    print_to_usb(&drawer_cmd)
}

/// Send ESC/POS command to open cash drawer via serial port.
pub fn open_drawer_serial(port_name: &str) -> Result<(), String> {
    let drawer_cmd = vec![0x1B, 0x70, 0x00, 0x19, 0xFA];
    print_to_serial(port_name, &drawer_cmd)
}

/// Send ESC/POS command to open cash drawer via network.
pub fn open_drawer_network(ip: &str, port: u16) -> Result<(), String> {
    let drawer_cmd = vec![0x1B, 0x70, 0x00, 0x19, 0xFA];
    print_to_network(ip, port, &drawer_cmd)
}

File: desktop/electron/printer.ts
import net from 'node:net';

export interface PrinterInfo {
  name: string;
  port_type: 'usb' | 'serial' | 'network';
  vendor_id: number | null;
  product_id: number | null;
  port_name: string | null;
  connected: boolean;
}

export async function listPrinters(): Promise<PrinterInfo[]> {
  const printers: PrinterInfo[] = [];

  // Enumerate serial ports
  try {
    const { SerialPort } = await import('serialport');
    const ports = await SerialPort.list();
    for (const port of ports) {
      if (port.path && (port.path.startsWith('COM') || port.path.startsWith('/dev/'))) {
        printers.push({
          name: port.path,
          port_type: 'serial',
          vendor_id: port.vendorId ? parseInt(port.vendorId, 16) : null,
          product_id: port.productId ? parseInt(port.productId, 16) : null,
          port_name: port.path,
          connected: true,
        });
      }
    }
  } catch (e) {
    console.warn('[Printer] serialport not available:', (e as Error).message);
  }

  // Enumerate USB printers
  try {
    const usb = await import('usb');
    const devices = usb.getDeviceList();
    for (const device of devices) {
      const desc = device.deviceDescriptor;
      // USB Printer class = 0x07
      if (device.configDescriptors?.[0]?.interfaces) {
        for (const iface of device.configDescriptors[0].interfaces) {
          for (const alt of iface) {
            if (alt.bInterfaceClass === 0x07) {
              printers.push({
                name: `USB Printer (VID:${desc.idVendor.toString(16)} PID:${desc.idProduct.toString(16)})`,
                port_type: 'usb',
                vendor_id: desc.idVendor,
                product_id: desc.idProduct,
                port_name: null,
                connected: true,
              });
            }
          }
        }
      }
    }
  } catch (e) {
    console.warn('[Printer] usb not available:', (e as Error).message);
  }

  return printers;
}

export async function printToPort(
  portType: string,
  portName: string,
  ip: string,
  networkPort: number,
  bytes: number[]
): Promise<void> {
  const buffer = Buffer.from(bytes);

  switch (portType) {
    case 'usb':
      await printUsb(buffer);
      break;
    case 'serial':
      await printSerial(portName, buffer);
      break;
    case 'network':
      await printNetwork(ip, networkPort, buffer);
      break;
    default:
      throw new Error(`Unknown printer type: ${portType}`);
  }
}

export async function openDrawerToPort(
  portType: string,
  portName: string,
  ip: string,
  networkPort: number
): Promise<void> {
  // ESC/POS pulse command for cash drawer pin 2
  const pulse = Buffer.from([0x1b, 0x70, 0x00, 0x19, 0xfa]);
  await printToPort(portType, portName, ip, networkPort, Array.from(pulse));
}

async function printUsb(buffer: Buffer): Promise<void> {
  try {
    const usb = await import('usb');
    const devices = usb.getDeviceList();

    for (const device of devices) {
      const desc = device.deviceDescriptor;
      if (!device.configDescriptors?.[0]?.interfaces) continue;

      for (const iface of device.configDescriptors[0].interfaces) {
        for (const alt of iface) {
          if (alt.bInterfaceClass !== 0x07) continue;

          try {
            device.open();
            device.selectConfiguration(1);
            device.claimInterface(alt.bInterfaceNumber);

            const endpoint = alt.endpoints?.find(
              (ep: any) => ep.bEndpointAddress & 0x80
            );
            if (!endpoint) continue;

            const chunkSize = 4096;
            for (let i = 0; i < buffer.length; i += chunkSize) {
              const chunk = buffer.subarray(i, Math.min(i + chunkSize, buffer.length));
              device.transferOut(endpoint.bEndpointAddress, chunk);
            }

            device.releaseInterface(alt.bInterfaceNumber);
            device.close();
            console.log('[Printer] USB print sent');
            return;
          } catch (e) {
            console.warn('[Printer] USB print attempt failed:', (e as Error).message);
            try { device.close(); } catch {}
          }
        }
      }
    }

    throw new Error('No USB printer found');
  } catch (e) {
    if ((e as Error).message === 'No USB printer found') throw e;
    throw new Error(`USB print failed: ${(e as Error).message}`);
  }
}

async function printSerial(portName: string, buffer: Buffer): Promise<void> {
  const { SerialPort } = await import('serialport');
  const { ReadlineParser } = await import('@serialport/parser-readline');

  return new Promise((resolve, reject) => {
    const port = new SerialPort({
      path: portName,
      baudRate: 9600,
    });

    const parser = port.pipe(new ReadlineParser());

    port.on('open', () => {
      port.write(buffer, (err) => {
        if (err) {
          port.close();
          reject(new Error(`Serial write failed: ${err.message}`));
        } else {
          setTimeout(() => {
            port.close();
            resolve();
          }, 500);
        }
      });
    });

    port.on('error', (err) => {
      reject(new Error(`Serial port error: ${err.message}`));
    });

    setTimeout(() => {
      port.close();
      reject(new Error('Serial print timeout'));
    }, 5000);
  });
}

async function printNetwork(ip: string, port: number, buffer: Buffer): Promise<void> {
  return new Promise((resolve, reject) => {
    const socket = new net.Socket();

    socket.setTimeout(5000);

    socket.connect(port, ip, () => {
      socket.write(buffer, (err) => {
        if (err) {
          socket.destroy();
          reject(new Error(`Network write failed: ${err.message}`));
        } else {
          setTimeout(() => {
            socket.destroy();
            resolve();
          }, 500);
        }
      });
    });

    socket.on('timeout', () => {
      socket.destroy();
      reject(new Error('Network print timeout'));
    });

    socket.on('error', (err) => {
      socket.destroy();
      reject(new Error(`Network print error: ${err.message}`));
    });
  });
}
